File: app/src/main/java/com/demo/networkstatsmanager/view/StatsActivity.java
package com.demo.networkstatsmanager.view;

import android.Manifest;
import android.annotation.TargetApi;
import android.app.AppOpsManager;
import android.app.usage.NetworkStatsManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.TrafficStats;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import com.bumptech.glide.Glide;
import com.demo.networkstatsmanager.R;
import com.demo.networkstatsmanager.model.Package;
import com.demo.networkstatsmanager.utils.NetworkStatsHelper;
import com.demo.networkstatsmanager.utils.PackageManagerHelper;

public class StatsActivity extends AppCompatActivity {
    private static final int READ_PHONE_STATE_REQUEST = 37;

    ImageView ivIcon;

    TextView trafficStatsPackageRx;
    TextView trafficStatsPackageTx;

    TextView networkStatsPackageRx;
    TextView networkStatsPackageTx;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_stats);
        ivIcon = findViewById(R.id.avatar);
    }

    @Override
    protected void onStart() {
        super.onStart();
        requestPermissions();
    }

    @Override
    @TargetApi(Build.VERSION_CODES.M)
    protected void onResume() {
        super.onResume();
        if (!hasPermissions()) {
            return;
        }
        Glide.with(this).load("https://en.wikipedia.org/wiki/McLaren_F1#/media/File:1996_McLaren_F1_Chassis_No_63_6.1_Front.jpg").into(ivIcon);

        initTextViews();
        intiData();
    }

    private void intiData() {
        String packageName = getPackageData().getPackageName();
        if (packageName == null) {
            return;
        }
        try {
            ivIcon.setImageDrawable(getPackageManager().getApplicationIcon(packageName));
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        if (!PackageManagerHelper.isPackage(StatsActivity.this, packageName)) {
            return;
        }
        fillData(packageName);
    }

    private void requestPermissions() {
        if (!hasPermissionToReadNetworkHistory()) {
            return;
        }
        if (!hasPermissionToReadPhoneStats()) {
            requestPhoneStateStats();
        }
    }

    private boolean hasPermissions() {
        return hasPermissionToReadNetworkHistory() && hasPermissionToReadPhoneStats();
    }

    private void initTextViews() {
        trafficStatsPackageRx = (TextView) findViewById(R.id.traffic_stats_package_rx_value);
        trafficStatsPackageTx = (TextView) findViewById(R.id.traffic_stats_package_tx_value);
        networkStatsPackageRx = (TextView) findViewById(R.id.network_stats_package_rx_value);
        networkStatsPackageTx = (TextView) findViewById(R.id.network_stats_package_tx_value);
    }

    private void fillData(String packageName) {
        int uid = PackageManagerHelper.getPackageUID_byName(this, packageName);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            NetworkStatsManager networkStatsManager = (NetworkStatsManager) getApplicationContext().getSystemService(Context.NETWORK_STATS_SERVICE);
            NetworkStatsHelper networkStatsHelper = new NetworkStatsHelper(networkStatsManager, uid);
            fillNetworkStatsPackage(uid, networkStatsHelper);
        }
        fillTrafficStatsPackage(uid);
    }

    @TargetApi(Build.VERSION_CODES.M)
    private void fillNetworkStatsPackage(int uid, NetworkStatsHelper networkStatsHelper) {

        long recievedRxKiloBytes = (networkStatsHelper.getPackageRxBytesMobile(uid, this) + networkStatsHelper.getPackageRxBytesWifi()) / 1024;
        long sentRxKiloBytes = (networkStatsHelper.getPackageTxBytesMobile(this) + networkStatsHelper.getPackageTxBytesWifi()) / 1024;

        if (recievedRxKiloBytes > 1) {
            networkStatsPackageRx.setText(recievedRxKiloBytes + " KBs recieved");
        } else {
            networkStatsPackageRx.setText(recievedRxKiloBytes + " KB recieved");
        }

        if (sentRxKiloBytes > 1) {
            networkStatsPackageTx.setText(sentRxKiloBytes + " KBs sent");
        } else {
            networkStatsPackageTx.setText(sentRxKiloBytes + " KB sent");
        }
    }

    private void fillTrafficStatsPackage(int uid) {
        long recievedTxKiloBytes = TrafficStats.getUidRxBytes(uid) / 1024;
        long sentTxKiloBytes = TrafficStats.getUidTxBytes(uid) / 1024;

        if (recievedTxKiloBytes > 1) {
            trafficStatsPackageRx.setText(recievedTxKiloBytes + " KBs recieved");
        } else {
            trafficStatsPackageRx.setText(recievedTxKiloBytes + " KB recieved");
        }

        if (sentTxKiloBytes > 1) {
            trafficStatsPackageTx.setText(sentTxKiloBytes + " KBs sent");
        } else {
            trafficStatsPackageTx.setText(sentTxKiloBytes + " KB sent");
        }
    }

    private boolean hasPermissionToReadPhoneStats() {
        return ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_DENIED;
    }

    private void requestPhoneStateStats() {
        ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_PHONE_STATE}, READ_PHONE_STATE_REQUEST);
    }

    private boolean hasPermissionToReadNetworkHistory() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            return true;
        }
        final AppOpsManager appOps = (AppOpsManager) getSystemService(Context.APP_OPS_SERVICE);
        int mode = appOps.checkOpNoThrow(AppOpsManager.OPSTR_GET_USAGE_STATS,
                android.os.Process.myUid(), getPackageName());
        if (mode == AppOpsManager.MODE_ALLOWED) {
            return true;
        }
        appOps.startWatchingMode(AppOpsManager.OPSTR_GET_USAGE_STATS,
                getApplicationContext().getPackageName(),
                new AppOpsManager.OnOpChangedListener() {
                    @Override
                    @TargetApi(Build.VERSION_CODES.M)
                    public void onOpChanged(String op, String packageName) {
                        int mode = appOps.checkOpNoThrow(AppOpsManager.OPSTR_GET_USAGE_STATS,
                                android.os.Process.myUid(), getPackageName());
                        if (mode != AppOpsManager.MODE_ALLOWED) {
                            return;
                        }
                        appOps.stopWatchingMode(this);
                        Intent intent = new Intent(StatsActivity.this, StatsActivity.class);
                        if (getIntent().getExtras() != null) {
                            intent.putExtras(getIntent().getExtras());
                        }
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                        getApplicationContext().startActivity(intent);
                    }
                });
        requestReadNetworkHistoryAccess();
        return false;
    }

    private void requestReadNetworkHistoryAccess() {
        Intent intent;
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
            startActivity(intent);
        }
    }

    private Package getPackageData() {
        PackageManager packageManager = getPackageManager();
        try {
            ApplicationInfo mApplicationInfo;
            PackageInfo packageInfo = getPackageManager().getPackageInfo(getPackageName(), 0);
            Package packageItem = new Package();
            packageItem.setVersion(packageInfo.versionName);
            packageItem.setPackageName(packageInfo.packageName);
            mApplicationInfo = packageManager.getApplicationInfo(packageInfo.packageName, PackageManager.GET_META_DATA);
            if (mApplicationInfo == null) {
                return null;
            }
            CharSequence appName = packageManager.getApplicationLabel(mApplicationInfo);
            if (appName != null) {
                packageItem.setName(appName.toString());
            }
            return packageItem;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return null;
        }

    }
}

File: app/src/main/java/com/demo/networkstatsmanager/view/StatsActivity.kt
package com.demo.networkstatsmanager.view

import android.Manifest
import com.demo.networkstatsmanager.utils.PackageManagerHelper.isPackage
import com.demo.networkstatsmanager.utils.PackageManagerHelper.getPackageUID_byName
import androidx.appcompat.app.AppCompatActivity
import android.widget.TextView
import android.os.Bundle
import com.demo.networkstatsmanager.R
import android.annotation.TargetApi
import android.os.Build
import com.bumptech.glide.Glide
import android.content.pm.PackageManager
import android.app.usage.NetworkStatsManager
import com.demo.networkstatsmanager.utils.NetworkStatsHelper
import android.net.TrafficStats
import androidx.core.app.ActivityCompat
import android.app.AppOpsManager
import android.app.AppOpsManager.OnOpChangedListener
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.os.Process
import android.provider.Settings
import android.view.View
import android.widget.ImageView
import com.demo.networkstatsmanager.model.Package

class StatsActivity : AppCompatActivity() {
  var ivIcon: ImageView? = null
  var trafficStatsPackageRx: TextView? = null
  var trafficStatsPackageTx: TextView? = null
  var networkStatsPackageRx: TextView? = null
  var networkStatsPackageTx: TextView? = null
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_stats)
    ivIcon = findViewById(R.id.avatar)
  }

  override fun onStart() {
    super.onStart()
    requestPermissions()
  }

  @TargetApi(Build.VERSION_CODES.M)
  override fun onResume() {
    super.onResume()
    if (!hasPermissions()) {
      return
    }
    Glide.with(this).load("https://en.wikipedia.org/wiki/McLaren_F1#/media/File:1996_McLaren_F1_Chassis_No_63_6.1_Front.jpg").into(ivIcon!!)
    initTextViews()
    intiData()
  }

  private fun intiData() {
    val packageName = packageData!!.packageName ?: return
    try {
      ivIcon!!.setImageDrawable(packageManager.getApplicationIcon(packageName))
    } catch (e: PackageManager.NameNotFoundException) {
      e.printStackTrace()
    }
    if (!isPackage(this@StatsActivity, packageName)) {
      return
    }
    fillData(packageName)
  }

  private fun requestPermissions() {
//        if (!hasPermissionToReadNetworkHistory()) {
//            return;
//        }
    if (!hasPermissionToReadPhoneStats()) {
      requestPhoneStateStats()
    }
  }

  //    private boolean hasPermissions() {
  //        return hasPermissionToReadNetworkHistory() && hasPermissionToReadPhoneStats();
  //    }
  private fun hasPermissions(): Boolean {
    return hasPermissionToReadPhoneStats()
  }

  private fun initTextViews() {
    trafficStatsPackageRx = findViewById<View>(R.id.traffic_stats_package_rx_value) as TextView
    trafficStatsPackageTx = findViewById<View>(R.id.traffic_stats_package_tx_value) as TextView
    networkStatsPackageRx = findViewById<View>(R.id.network_stats_package_rx_value) as TextView
    networkStatsPackageTx = findViewById<View>(R.id.network_stats_package_tx_value) as TextView
  }

  private fun fillData(packageName: String) {
    val uid = getPackageUID_byName(this, packageName)
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      val networkStatsManager = applicationContext.getSystemService(NETWORK_STATS_SERVICE) as NetworkStatsManager
      val networkStatsHelper = NetworkStatsHelper(networkStatsManager, uid)
      fillNetworkStatsPackage(uid, networkStatsHelper)
    }
    fillTrafficStatsPackage(uid)
  }

  @TargetApi(Build.VERSION_CODES.M)
  private fun fillNetworkStatsPackage(uid: Int, networkStatsHelper: NetworkStatsHelper) {
    val recievedRxKiloBytes = (networkStatsHelper.getPackageRxBytesMobile(this) + networkStatsHelper.packageRxBytesWifi) / 1024
    val sentRxKiloBytes = (networkStatsHelper.getPackageTxBytesMobile(this) + networkStatsHelper.packageTxBytesWifi) / 1024
    if (recievedRxKiloBytes > 1) {
      networkStatsPackageRx!!.text = "$recievedRxKiloBytes KBs recieved"
    } else {
      networkStatsPackageRx!!.text = "$recievedRxKiloBytes KB recieved"
    }
    if (sentRxKiloBytes > 1) {
      networkStatsPackageTx!!.text = "$sentRxKiloBytes KBs sent"
    } else {
      networkStatsPackageTx!!.text = "$sentRxKiloBytes KB sent"
    }
  }

  private fun fillTrafficStatsPackage(uid: Int) {
    val recievedTxKiloBytes = TrafficStats.getUidRxBytes(uid) / 1024
    val sentTxKiloBytes = TrafficStats.getUidTxBytes(uid) / 1024
    if (recievedTxKiloBytes > 1) {
      trafficStatsPackageRx!!.text = "$recievedTxKiloBytes KBs recieved"
    } else {
      trafficStatsPackageRx!!.text = "$recievedTxKiloBytes KB recieved"
    }
    if (sentTxKiloBytes > 1) {
      trafficStatsPackageTx!!.text = "$sentTxKiloBytes KBs sent"
    } else {
      trafficStatsPackageTx!!.text = "$sentTxKiloBytes KB sent"
    }
  }

  private fun hasPermissionToReadPhoneStats(): Boolean {
    return ActivityCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE) != PackageManager.PERMISSION_DENIED
  }

  private fun requestPhoneStateStats() {
    ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_PHONE_STATE), READ_PHONE_STATE_REQUEST)
  }

  private fun hasPermissionToReadNetworkHistory(): Boolean {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
      return true
    }
    val appOps = getSystemService(APP_OPS_SERVICE) as AppOpsManager
    val mode = appOps.checkOpNoThrow(
      AppOpsManager.OPSTR_GET_USAGE_STATS,
      Process.myUid(), packageName
    )
    if (mode == AppOpsManager.MODE_ALLOWED) {
      return true
    }
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
      appOps.startWatchingMode(AppOpsManager.OPSTR_GET_USAGE_STATS,
        applicationContext.packageName,
        object : OnOpChangedListener {
          @TargetApi(Build.VERSION_CODES.M)
          override fun onOpChanged(op: String, packageName: String) {
            val mode = appOps.checkOpNoThrow(
              AppOpsManager.OPSTR_GET_USAGE_STATS,
              Process.myUid(), getPackageName()
            )
            if (mode != AppOpsManager.MODE_ALLOWED) {
              return
            }
            appOps.stopWatchingMode(this)
            val intent = Intent(this@StatsActivity, StatsActivity::class.java)
            if (getIntent().extras != null) {
              intent.putExtras(getIntent().extras!!)
            }
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
            applicationContext.startActivity(intent)
          }
        })
    }
    requestReadNetworkHistoryAccess()
    return false
  }

  private fun requestReadNetworkHistoryAccess() {
    val intent: Intent
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
      startActivity(intent)
    }
  }

  private val packageData: Package?
    private get() {
      val packageManager = packageManager
      return try {
        val mApplicationInfo: ApplicationInfo
        val packageInfo = getPackageManager().getPackageInfo(packageName, 0)
        val packageItem = Package()
        packageItem.version = packageInfo.versionName
        packageItem.packageName = packageInfo.packageName
        mApplicationInfo = packageManager.getApplicationInfo(packageInfo.packageName, PackageManager.GET_META_DATA)
        if (mApplicationInfo == null) {
          return null
        }
        val appName = packageManager.getApplicationLabel(mApplicationInfo)
        if (appName != null) {
          packageItem.name = appName.toString()
        }
        packageItem
      } catch (e: PackageManager.NameNotFoundException) {
        e.printStackTrace()
        null
      }
    }

  companion object {
    private const val READ_PHONE_STATE_REQUEST = 37
  }
}